docs: Custom rule & plugin tutorial (#17024)
eslint/eslint: docs/src/extend/custom-rule-plugin-tutorial.md

Author: `docs/src/extend/custom-rule-plugin-tutorial.md` is the correct tutorial. updated the PR with just this.

sorry for confusion.

---

> title: Custom Rule & Plugin Tutorial

Reviewer: I think we may just want to call this "Custom Rule Tutorial". I re-read this a couple of times and I kept thinking it reads like two separate tutorials will be in this page. Logically, creating a plugin is the latest step in creating a custom rule, so I think we're okay with shortening this title.
```suggestion
title: Custom Rule Tutorial
```

Author: will also update the page name

eleventyNavigation:
    key: custom rule plugin tutorial
    parent: create plugins
    title: Custom Rule & Plugin Tutorial
    order: 1
---

> This tutorial covers how to create a custom rule and plugin for ESLint.

Reviewer: ```suggestion
This tutorial covers how to create a custom rule for ESLint and distribute it with a plugin.
```

You can create custom rules to validate if your code meets a certain expectation, and determine what to do if it does not meet that expectation. Plugins package custom rules and other configuration, allowing you to easily share and reuse them in different projects.

To learn more about custom rules and plugins refer to the following documentation:

* [Custom Rules](custom-rules)
* [Plugins](plugins)

## Why Create a Custom Rule?

> Create a custom rule if the ESLint [built-in rules](../rules/) and community-published custom rules do not meet your needs.

Reviewer: Let's expand on this a bit.

```suggestion
Create a custom rule if the ESLint [built-in rules](../rules/) and community-published custom rules do not meet your needs. You might create a custom rule to enforce a best practice for your company or project, to prevent a particular bug from recurring, or to ensure compliance with a style guide.   
```

> Before creating a custom rule, it's worth investigating on the web if someone has published a plugin with a custom rule that solves your use case. Very often the rule will already exist.

Reviewer: ```suggestion
Before creating a custom rule that isn't specific to your company or project, it's worth searching the web to see someone has published a plugin with a custom rule that solves your use case. It's quite possible the rule may already exist.
```

```shell
mkdir foo-bar-rule # create directory
cd foo-bar-rule # enter the directory
npm init -y # init new npm project
touch foo-bar.js # create file foo-bar.js
```

## Step 2: Stub Out Rule File

In the `foo-bar.js` file, add some scaffolding for the `foo-bar` custom rule. Also add a `meta` object with some basic information about the rule.

All ESLint rules are objects that follow this structure.

```javascript
// foo-bar.js

module.exports = {
  meta: {
     // TODO: add metadata
    },
    create(context) {
        return {
            // TODO: add callback function(s)
        };
    }
};
```

## Step 3: Add Rule Metadata

Before writing the rule, add some metadata to the rule object. ESLint uses this information when running the rule.

Start by exporting a module with a `meta` object containing the rule's metadata, such as the rule type, documentation, and fixability. In this case, the rule type is "fix," the description is "Can only assign 'bar' to const foo.", and the rule is fixable by modifying the code.

```javascript
// foo-bar.js

module.exports = {
    meta: {

>         type: "fix",

Reviewer: "fix" isn't a valid option for `type`. It must be "problem", "suggestion", or "layout"

```suggestion
        type: "problem",
```

Author: note to self: fix in the source code as well

        docs: {
            description: "Can only assign 'bar' to `const foo`.",
        },
        fixable: "code"
    },
    create(context) {
        return {
            // TODO: add callback function(s)
            }
        };
    }
};
```

To learn more about rule metadata, refer to [Rule Structure](custom-rules#rule-structure).

> ## Step 4: Add Rule Callback Functions

Reviewer: I'm not sure what "callback functions" refers to here. I think maybe you're referring to the `VariableDeclaration` method? If that's the case, I would suggest instead calling this a visitor method.

```suggestion
## Step 4: Add Rule Visitor Methods
```

Define the rule's `create` function, which accepts a `context` object and returns an object with a property for each syntax node type you want to handle. In this case, we want to handle `"VariableDeclaration"` nodes.
You can choose any [ESTree node type](https://github.com/estree/estree) or [selector](selectors).

> Inside the `"VariableDeclaration"` function, check if the node represents a `const` variable declaration, if its name is `foo`, and if it's not assigned to the string `"bar"`. You do this by evaluating the `node` passed to the `"VariableDeclaration"` function.

Reviewer: ```suggestion
Inside the `"VariableDeclaration"` visitor method, check if the node represents a `const` variable declaration, if its name is `foo`, and if it's not assigned to the string `"bar"`. You do this by evaluating the `node` passed to the `"VariableDeclaration"` method.
```

If the `const foo` declaration_is assigned to `"bar"` the rule does nothing. If `const foo` **is not** assigned to `"bar"`, then `context.report()` reports an error to ESLint. The error report include information about the error and how to fix it.

```javascript
// foo-bar.js
module.exports = {
    meta: {
        type: "fix",
        docs: {
            description: "Can only assign 'bar' to `const foo`.",
        },
         fixable: "code"
    },
    create(context) {
        return {
            // Performs action in the function on every variable declaration

>             "VariableDeclaration": function(node) {

Reviewer: Let's use the modern syntax
```suggestion
            VariableDeclaration(node) {
```

                            // Report error to ESLint. Error message uses
                            // a message placeholder to include the incorrect value
                            // in the error message.
                            // Also includes a `fix(fixer)` function that replaces
                            // any values assigned to `const foo` with "bar".
                            context.report({
                                node,
                                message: 'Value other than "bar" assigned to \`const foo\`. Unexpected value: {{ notBar }}',
                                data: {
                                    notBar: node.declaration.init.value
                                },
                                fix(fixer) {
                                    return fixer.replaceText(node.declaration.init, "bar");
                                }
                            });
                        }
                    }
                }
            }
        };
    }
};
```

## Step 5: Set up Testing

With the rule written, you can test it to make sure it's working as expected.

> ESLint provides the built-in [RuleTester](../integrate/nodejs-api#ruletester) class to test rules. You do not need to use 3rd-party testing libraries to test ESLint rules.

Reviewer: ```suggestion
ESLint provides the built-in [`RuleTester`](../integrate/nodejs-api#ruletester) class to test rules. You do not need to use third-party testing libraries to test ESLint rules, but `RuleTester` works seamlessly with tools like Mocha and Jest.
```

  parserOptions: { ecmaVersion: 2015 }
});

// Throws error if the tests in ruleTester.run() do not pass
ruleTester.run(
  "foo-bar", // rule name
  fooBarRule, // rule code
  { // checks
    // 'valid' checks cases that should pass
    valid: [{
      code: "const foo = 'bar';",
    }],
    // 'invalid' checks cases that should not pass
    invalid: [{
      code:"const foo = 'baz';",
      output: 'const foo = "bar";',
      errors: 1,
    }],
  }
);

console.log("All tests passed!");
```

## Step 7: Bundle the Custom Rule in a Plugin

Now that you've written the custom rule and validated that it works, you can include it in a plugin. Using a plugin, you can share the rule in an npm package to use in other projects.

Create the file for the plugin:

```shell
touch eslint-plugin-foo-bar.js
```

And now write the plugin code. Plugins are just exported JavaScript objects. To include a rule in a plugin, include it in the plugin's `rules` object, which contains key-value pairs of rule names and their source code.

To learn more about creating plugins, refer to [Create Plugins](plugins).

```javascript
// eslint-plugin-foo-bar.js

const fooBarRule = require("./foo-bar");

> module.exports = { rules: { "foo-bar": fooBarRule } };

Reviewer: I'd suggest breaking this out into multiple lines for easier reading.

```

## Step 8: Publish the Plugin

To publish a plugin containing a rule to npm, you need to do a bit of set up in your project's `package.json` file.

> To help publish the package easily, install the dependency [np](https://www.npmjs.com/package/np):

Reviewer: This seems like an unnecessary step? Why not just to do `npm publish`?

```shell
npm install --save-dev np
```

Add a script to run `np`:

```javascript
// package.json
// ...rest of package.json
  "scripts": {
    "test": "node foo-bar.test.js",
    "pub": "np --yarn false" // new publication script
  },
// ...rest of package.json
```

Next, configure the `package.json` to be ready for publishing the plugin. Add the following in the corresponding fields:

1. `"name"`: A unique name for the package. No other package on npm can have the same name.
1. `"main"`: The relative path to the plugin file. Following this example, the path is `"eslint-plugin-foo-bar.js"`.
1. `"description"`: A description of the package that's viewable on npm.

> 1. `"peerDependencies"`: Add `"eslint": ">=0.8.0"` as a peer dependency. Plugins were introduced in eslint@0.8.0, so any version greater than or equal to that is necessary to use the plugin. By declaring `eslint` as a peer dependency, it requires that users add the package to the project separately from the plugin.

Reviewer: We may want to use `>=8.0.0` here. It's unlikely newer rules will work all the way back to 0.8.0.

Author: note to self: update in src code too

A complete annotated example of what a plugin's `package.json` file should look like:

```javascript
// package.json
{
  // name npm package.
  // Add your own package name. eslint-plugin-foo-bar is taken!
  "name": "eslint-plugin-foo-bar",
  "version": "1.2.0",
  "description": "ESLint plugin for foo-bar rule.",
  "main": "eslint-plugin-foo-bar.js", // plugin entry point
  "scripts": {
    "test": "node foo-bar.test.js",
    "pub": "np --yarn false"
  },
  // Add eslint>=0.8.0 as a peer dependency.
  // This was the ESLint version where plugins were introduced.
  "peerDependencies": {
    "eslint": ">=0.8.0"
  },
  // Add these standard keywords to make plugin easy to find!
  "keywords": [
    "eslint",
    "eslintplugin",
    "eslint-plugin"
  ],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "eslint": "^8.36.0",
    "np": "^7.6.3"
  }
}
```

To publish the package, run `npm run pub` and follow the CLI prompts.

You should see the package live on npm!

## Step 9: Use the Published Custom Rule

To use the package now that it's published, run the following command in your project to download your published package:

```shell

> npm install --save-dev <YOUR-PLUGIN-PACKAGE-NAME>

Reviewer: Let's use the actual plugin name here to make it easier to follow the tutorial.

```

## Step 2: Stub Out Rule File

In the `foo-bar.js` file, add some scaffolding for the `foo-bar` custom rule. Also add a `meta` object with some basic information about the rule.

All ESLint rules are objects that follow this structure.

```javascript
// foo-bar.js

module.exports = {
  meta: {
     // TODO: add metadata
    },
    create(context) {
        return {
            // TODO: add callback function(s)
        };
    }
};
```

## Step 3: Add Rule Metadata

Before writing the rule, add some metadata to the rule object. ESLint uses this information when running the rule.

Start by exporting a module with a `meta` object containing the rule's metadata, such as the rule type, documentation, and fixability. In this case, the rule type is "fix," the description is "Can only assign 'bar' to const foo.", and the rule is fixable by modifying the code.

```javascript
// foo-bar.js

module.exports = {
    meta: {
        type: "fix",
        docs: {
            description: "Can only assign 'bar' to `const foo`.",
        },
        fixable: "code"
    },
    create(context) {
        return {
            // TODO: add callback function(s)
            }
        };
    }
};
```

> To learn more about rule metadata, refer to [Rule Structure](custom-rules#rule-structure).

Author: note this link is causing the CI to fail here https://github.com/eslint/eslint/actions/runs/4549420021/jobs/8021476473?pr=17024

that's b/c this link uses the internal page link that's to be added in https://github.com/eslint/eslint/pull/16906/files#diff-dc3f44ef800795b890a57c8dd95c6ce26ba8a47a4551d7a05d7a5cd9cf016597R37

Reviewer: No worries. We'll merge that PR first and then we can rebase this one.
